pkg/common-controller/snapshot_controller.go

```diff
@@ -1392,24 +1418,53 @@ func isControllerUpdateFailError(err *crdv1.VolumeSnapshotError) bool {
 
 // addSnapshotFinalizer adds a Finalizer for VolumeSnapshot.
 func (ctrl *csiSnapshotCommonController) addSnapshotFinalizer(snapshot *crdv1.VolumeSnapshot, addSourceFinalizer bool, addBoundFinalizer bool) error {
-	snapshotClone := snapshot.DeepCopy()
-	if addSourceFinalizer {
-		snapshotClone.ObjectMeta.Finalizers = append(snapshotClone.ObjectMeta.Finalizers, utils.VolumeSnapshotAsSourceFinalizer)
-	}
-	if addBoundFinalizer {
-		snapshotClone.ObjectMeta.Finalizers = append(snapshotClone.ObjectMeta.Finalizers, utils.VolumeSnapshotBoundFinalizer)
-	}
-	newSnapshot, err := ctrl.clientset.SnapshotV1().VolumeSnapshots(snapshotClone.Namespace).Update(context.TODO(), snapshotClone, metav1.UpdateOptions{})
-	if err != nil {
-		return newControllerUpdateError(utils.SnapshotKey(snapshot), err.Error())
+	var updatedSnapshot *crdv1.VolumeSnapshot
+	var err error
+
+	// Must perform an update if no finalizers exist
```

Comment:
Do you know why we must perform an update if no finalizers exist?

Reply:
good catch, this was from before I found out about the "replace" patch operation with a new array. We can probably do that here instead. I did this with the content finalizers: https://github.com/kubernetes-csi/external-snapshotter/pull/526/files/0ccf8017800f0f6160f5003f2cf333954dfa4a91#diff-1a3473f9016f7abcf5007414ff7141e39032aaf7aba079eeef262d04167208a2R813-R830

Reply:
I just tried this, but patch seems to keep failing for this case. It's pretty hard to debug/find out why, so I would prefer to keep this as Update unless we start seeing errors related to this section. I just updated the PR to reflect this.

I've added a unit test for the patch case where we add a single finalizer, and that seems to work.

Reply:
added this